[SDK/CLI] added autocompletion for pf cli

sashokbg opened the pull request:
# Description

This pull request adds shell (bash / zsh) autocompletion for the "pf" cli tool.
It is based on "argcomplete" library https://kislyuk.github.io/argcomplete/ and inspired by "az" cli 

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

sashokbg commented:
Hello @wangchao1230 I created this PR as a draft related to #2446 
Can you please tell me if I am on the right track before I continue with the rest it ?

Thank you and have a nice day

wangchao1230 commented:
> Hello @wangchao1230 I created this PR as a draft related to #2446 Can you please tell me if I am on the right track before I continue with the rest it ?
> 
> Thank you and have a nice day

I think we are on the right track. However we are doing a major refactor to layering our sdk into several packages recently. And which has caused some conflict to your PR. 

+ @YingChen1996 to help you on that, and help you merge to main finally

wangchao1230 commented on src/promptflow/promptflow/_cli/_completers/_param_completers.py:
```diff
@@ -0,0 +1,19 @@
+from argcomplete import warn
```
please add more PR description on what features you tring to add in this PR

wangchao1230 commented on src/promptflow/promptflow/_cli/_completers/_param_completers.py:
```diff
@@ -0,0 +1,19 @@
+
+def run_name_completer(prefix, parsed_args, **kwargs):
+    client = PFClient()
+    runs = client.runs.list(max_results=50, list_view_type=ListViewType.ALL)
```
maybe add a constants for the magic number 50. will 50 be too much? 
I am not sure how will this looks if we have 50 results in the terminal?

wangchao1230 commented on autocomplete.sh:
```diff
@@ -0,0 +1,11 @@
+case $SHELL in
```
may move this to promptflow-devkit package

sashokbg commented:
> @sashokbg please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

sashokbg commented:
Hello @wangchao1230 I have rebased the MR and added some description per your comments.

@YingChen1996 can you please help me with how the pf.autocomplete.sh needs to be sourced ? 
Do we need to include it in the final "distribution" alongside the "pf" executable ?
Do we need to update the .zshhrc file of the client ?

Thank you for your help

wangchao1230 commented on src/promptflow-devkit/pyproject.toml:
```diff
@@ -68,6 +68,7 @@ pyinstaller = ">=5.13.2"
 streamlit = ">=1.26.0"
 streamlit-quill = "<0.1.0"
 bs4 = "*"
+argcomplete = ">=3.2.3"
```
add comment on why this deps is needed

YingChen1996 commented:
Hi, @sashokbg .
1. Yeah, I think user need add `source pf.autocomplete.sh` in their ~/.bashrc, ~/.zshrc or ~/.bash_profile.  Maybe we need add this note in the document.
2. Yes, we need include it in the final "distribution" alongside the "pf" executable.
3. What do you mean update the .zshhrc file of the client? I think it's likely similar to the first question, user need udpate  .zshhrc file to source pf.autocomplete.sh in zshell

YingChen1996 commented:
> Hello @wangchao1230 I have rebased the MR and added some description per your comments.
> 
> @YingChen1996 can you please help me with how the pf.autocomplete.sh needs to be sourced ? Do we need to include it in the final "distribution" alongside the "pf" executable ? Do we need to update the .zshhrc file of the client ?
> 
> Thank you for your help

Hi, @sashokbg .
1. Yeah, I think user need add `source pf.autocomplete.sh` in their ~/.bashrc, ~/.zshrc or ~/.bash_profile.  Maybe we need add this note in the document.
2. Yes, we need include it in the final "distribution" alongside the "pf" executable.
3. What do you mean update the .zshhrc file of the client? I think it's likely similar to the first question, user need udpate  .zshhrc file to source pf.autocomplete.sh in zshell

sashokbg commented:
> > Hello @wangchao1230 I have rebased the MR and added some description per your comments.
> > @YingChen1996 can you please help me with how the pf.autocomplete.sh needs to be sourced ? Do we need to include it in the final "distribution" alongside the "pf" executable ? Do we need to update the .zshhrc file of the client ?
> > Thank you for your help
> 
> Hi, @sashokbg .
> 
> 1. Yeah, I think user need add `source pf.autocomplete.sh` in their ~/.bashrc, ~/.zshrc or ~/.bash_profile.  Maybe we need add this note in the document.
> 2. Yes, we need include it in the final "distribution" alongside the "pf" executable.
> 3. What do you mean update the .zshhrc file of the client? I think it's likely similar to the first question, user need udpate  .zshhrc file to source pf.autocomplete.sh in zshell

Hello @YingChen1996 ok I will take care of the rest.
Can you point to me how do we include a file in the final distribution please ?

wangchao1230 commented on docs/reference/pf-command-reference.md:
```diff
@@ -4,6 +4,14 @@
 This is an experimental feature, and may change at any time. Learn [more](../how-to-guides/faq.md#stable-vs-experimental).
 :::
 
```
may put this at the end of the doc

wangchao1230 commented on src/promptflow-devkit/promptflow/_cli/_completers/_param_completers.py:
```diff
@@ -0,0 +1,21 @@
+
+def run_name_completer(prefix, parsed_args, **kwargs):
+    client = PFClient()
+    runs = client.runs.list(list_view_type=ListViewType.ALL)
```
do a max_results=MAX_RESULTS +1 here? So we won't query too much rows from the db in memory.

sashokbg replied:
Hello, I ended up by writing a new "search" method in the client that will search by name prefix.
I removed the "warn" for too many results since it breaks the view for some reason -> the method is now much simpler

wangchao1230 commented on src/promptflow-devkit/promptflow/_cli/_completers/_param_completers.py:
```diff
@@ -0,0 +1,21 @@
+
+    res = []
+
+    for entity in runs:
```
only return MAX_RESULTS of data?

wangchao1230 commented on src/promptflow/setup.py:
```diff
@@ -64,6 +64,7 @@
             "promptflow-core[executor-service]",  # used to build web executor server
         ],
     },
+    scripts=["pf.autocomplete.sh"],
```
@YingChen1996  Shall we inlcude this in devit setup?

crazygao replied:
Please include in devkit setup, by the way, pyproject.toml doesn't have a perfect way for this script.
Please insert something into rc file, just like what conda did.

YingChen1996 replied:
I think no need put it in setup.  

Maybe put pf.autocomplete.sh in the folder src/promptflow-devkit/promptflow is enough since it's for pf command and add  `source <promptflow_package_install_root>/pf.completion.sh` in doc.

sashokbg replied:
Hello everyone, I am sorry I am a little lost since python is not one of my strongs. Where should I put the .sh file and in what setup.py file should I reference it ?

Also how do I include it in the rc file ? Conda seems to update $PATH but I do not see how to execute my pf.autocomplete.sh on env activate

wangchao1230 replied:
@sashokbg Do you mind us make some changes to this PR? 

I think the major function is ready in this PR. And we can tweak some minor details and get this merged. 
Please let us know if this works for you.

sashokbg replied:
Hello @wangchao1230 yes no problem at all. How would you like to proceed, should I stop pushing to this branch ?

wangchao1230 commented on src/promptflow/pf.autocomplete.sh:
```diff
@@ -0,0 +1,11 @@
+case $SHELL in
```
We should add a changelog for this feature.

sashokbg replied:
Added in src/promptflow/CHANGELOG.md in my last commit

crazygao commented on docs/reference/pf-command-reference.md:
```diff
@@ -4,6 +4,14 @@
 
+## Autocomplete
+
+To activate autocomplete features for the pf CLI you need to add the following snippet to your ~/.bashrc or ~/.zshrc:
```
direct copy this into rc file?

sashokbg replied:
You mean "echo and append" it to the rc ? The problem is that we need to add logic that detects where your rc is and how it is called. In an XDG desktop it will be under ~/.config and in non XDG it will be under ~

Another issue is that if we have different "venvs" and have "pf" installed multiple times, which autocomplete should be used ?

wangchao1230 commented on src/promptflow-devkit/promptflow/_sdk/operations/_run_operations.py:
```diff
@@ -65,6 +65,29 @@ def list(
         )
 
+    @monitor_operation(activity_name="pf.runs.search", activity_type=ActivityType.INTERNALCALL)
+    def search(
```
Let's make this private now. 
Before we make this public:
- We need to go through some internal review to. 
- And we need to also make the pfazure has parity feature.

sashokbg replied:
Hello @wangchao1230  I thought that putting ```activity_type=ActivityType.INTERNALCALL``` was enough. Should I remove the @monitor_operation completely ?

wangchao1230 replied:
It will appear in reference doc if this is public fucnction: https://microsoft.github.io/promptflow/reference/python-library-reference/promptflow-devkit/promptflow.operations.html#promptflow.operations.RunOperations

It's just rename search() to _search()

sashokbg replied:
Done

wangchao1230 commented on src/promptflow/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 - [SDK/CLI] Create a run with `resume_from`, note that only run created with `promptflow>=1.8.0` can be used as the value of `resume_from`:
   - CLI: Support `pf run create --resume-from <original-run-name>` to create a run resume from another run.
   - SDK: Support `pf.run(resume_from=<original-run-name>)` to create a run resume from another run.
+  - CLI: Added autocomplete feature.
```
This should be in 1.9.0

wangchao1230 commented on src/promptflow/CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@
 - [SDK/CLI] Create a run with `resume_from`, note that only run created with `promptflow>=1.8.0` can be used as the value of `resume_from`:
   - CLI: Support `pf run create --resume-from <original-run-name>` to create a run resume from another run.
   - SDK: Support `pf.run(resume_from=<original-run-name>)` to create a run resume from another run.
+  - CLI: Added autocomplete feature.
```
This should be in 1.9.0  and include in both dekvit and promptflow

wangchao1230 commented on docs/reference/pf-command-reference.md:
```diff
@@ -897,4 +897,12 @@ Upgrade prompt flow without prompt and run non-interactively.
+
+## Autocomplete
+
+To activate autocomplete features for the pf CLI you need to add the following snippet to your ~/.bashrc or ~/.zshrc:
```
may need to include the version

YingChen1996 replied:
will fix in the next pr

wangchao1230 commented:
Thanks @sashokbg for your contribution! I have merged the PR. 
You can expect this changes goes in 1.9.0, which should come next week if goes well.
